src/transform/src/demand.rs

```diff
@@ -365,4 +366,65 @@ impl Demand {
             }
         })
     }
+
+    /// Checks the optimizer invariant that there are no more opportunities for projection pushdown,
+    /// comparing against Demand.
+    ///
+    /// More specifically, we check that the MFPs directly on top of global Gets project at most
+    /// those columns that `Demand` determines are demanded from that global Get. An exception is
+    /// when there is an `ArrangeBy` on top of a global `Get`, in which case the relevant MFP might
+    /// have been lifted away by `JoinImplementation` to enable index reuse. We skip checking the
+    /// MFP in this case.
+    ///
+    /// This is meant to be called at the end of the MIR pipeline, where it can catch 3 types of
+    /// issues:
+    /// 1. If `ProjectionPushdown` is not as smart as `Demand`.
+    /// 2. If some transform after the last run of `ProjectionPushdown` undoes something that
+    ///    `ProjectionPushdown` did.
+    /// 3. If some transform after the last run of `ProjectionPushdown` opens up new opportunities
+    ///    for projection pushdown.
+    pub fn soft_assert_no_more_projection_pushdown(
+        relation: &MirRelationExpr,
+    ) -> Result<(), crate::TransformError> {
+        let mut relation = relation.clone();
+        let arity = relation.arity();
+        // Gather demanded columns of each of the Ids.
+        let mut demand_on_ids = BTreeMap::new();
+        let demand = Demand::default();
+        demand.action(&mut relation, (0..arity).collect(), &mut demand_on_ids)?;
+        // Check that MFPs on top of global Gets project at most those columns that Demand thinks
+        // are demanded.
+        //
+        // We use `visit_pre_post`, because we'd like to control how we descend to children: we want
+        // to jump over an MFP that we find at the current node, in order to avoid observing partial
+        // MFPs, that are without their Projects.
+        relation.visit_pre_post(&mut |expr| {
+            let (mfp, expr) = MapFilterProject::extract_from_expression(expr);
+            match expr {
+                MirRelationExpr::Get { id: Id::Global(id), .. } => {
+                    let demand = demand_on_ids.get(&Id::Global(*id)).expect("`Demand` should have an opinion on all ids");
+                    let actual_proj_above_get = mfp.projection.iter().filter(|c| **c < mfp.input_arity).collect_vec();
+                    soft_assert_or_log!(
```

Comment:
Should `relation.pretty()` and `expr.pretty()` and `mfp` be redacted?

Reply:
Good point!
- `pretty` already redacts in prod: `pretty` calls `ExplainConfig::default()`, which has
`redacted: !mz_ore::assert::soft_assertions_enabled()`.
- `Display for MapFilterProject` unfortunately doesn't redact, and would be hard to make it do so properly. But since `Display for MapFilterProject` is only used in debugging code, I've now tweaked it to just omit the whole MFP if `!soft_assertions_enabled()`.